app/services/hts_service/reports/hts_report_builder.rb

```diff
@@ -38,115 +38,95 @@ def his_patients_rev
-            patient_obs = {}
-        
-            columns.each_with_index do |value, index|
-              patient_obs[value] = observations[index]
-            end
-        
-            patient_data = data[patient_id]
-            if patient_data.nil?
-              patient_data = { "patientid" => patient_id, "observations" => [], "demographics" => { "gender" => row[1], "person_id" => patient_id, "birthdate" => row[2], "encounter_datetime" => row[3] } }
-              data[patient_id] = patient_data
-            end
+      def his_patients_revs(indicators) 
+    
+        columns = ActiveRecord::Base.connection.columns('obs').map(&:name) 
+        sql_query = <<~SQL
+          SELECT
+            person.gender,
+            person.birthdate,
+            encounter.encounter_datetime,
+            obs.*
+          FROM patient
+          INNER JOIN person ON person.person_id = patient.patient_id
+            AND patient.voided = 0
+          INNER JOIN encounter ON encounter.patient_id = patient.patient_id
+            AND encounter.encounter_datetime BETWEEN '#{@start_date}' AND '#{@end_date}'
+            AND encounter.encounter_type = (SELECT encounter_type_id FROM encounter_type WHERE name = 'Testing')
+            AND encounter.voided = 0
+          INNER JOIN program ON program.program_id = encounter.program_id
+            AND program.program_id = (SELECT program_id FROM program WHERE name = 'HTC PROGRAM')
+            AND program.retired = 0
+          INNER JOIN obs ON obs.person_id = patient.patient_id
+            AND obs.voided = 0;
+        SQL
         
-            patient_data["observations"] << patient_obs
-          end
-        
-          return process_patient_data(data, indicators)
-    end
-        
-    def process_patient_data(data, indicators)
-          patient_data = []
+        results = ActiveRecord::Base.connection.select_all(sql_query)        
+         return process_patient_data(results, indicators)
+            
+      end
+            
+      def process_patient_data(results, indicators)  
+    
+        data = []
+    
+        grouped_obs = results.group_by { |obs| obs['person_id'] }
         
-          data.each do |_patient_id, patient_observation|
-            observations = patient_observation["observations"]
-            patient_obs = {}
+        grouped_obs.each_with_index do |row, index|
+          patient_id = row[0]
+          observations = row[1]    
+          data << { "person_id" => patient_id,"gender"=> observations[0]["gender"],"birthdate"=>observations[0]['birthdate'],"encounter_datetime"=>observations[0]["encounter_datetime"] }    
+          index_new = data.index(data.last)     
         
-            indicators.each do |indicator|
-              if indicator[:concept_id].is_a?(Integer)
-                desired_observation = observations.find { |obs| obs["concept_id"] == indicator[:concept_id].to_s }
+          indicators.each do |indicator|
```

Comment:
Can you replace this loop with this
```suggestion
indicators.each do |indicator|
    indicator_array = indicator[:concept_id].is_a?(Array) ? indicator[:concept_id] : [indicator[:concept_id]]
    indicator_values = indicator[:value]
    indicator_names = indicator[:name]
    indicator_array.each_with_index do |concept_id, index|
        desired_observation = observations.find { |obs| obs['concept_id'] == concept_id }
        name = indicator_names[index]
        data[index_new][name] = desired_observation&.[](indicator_values) if desired_observation
    end
end
```